cdap-app-fabric/src/main/java/co/cask/cdap/internal/provision/ProvisioningService.java

```diff
@@ -286,6 +339,44 @@ public void validateProperties(String provisionerName, Map<String, String> prope
     provisioner.validateProperties(properties);
   }
 
+  private Runnable createProvisionTask(ProvisioningTaskInfo taskInfo, Provisioner provisioner) {
+    ProgramRunId programRunId = taskInfo.getProgramRunId();
+    ProvisionerContext context = new DefaultProvisionerContext(programRunId, taskInfo.getProvisionerProperties(),
+                                                               sparkCompat, createSSHContext(taskInfo.getSshKeyInfo()));
+
+    // TODO: (CDAP-13246) pick up timeout from profile instead of hardcoding
+    return new ProvisionTask(taskInfo, provisioner, context, provisionerNotifier, transactional, datasetFramework, 300);
+  }
+
+  private Runnable createDeprovisionTask(ProvisioningTaskInfo taskInfo, Provisioner provisioner,
+                                         Consumer<ProgramRunId> taskCleanup) {
+    Map<String, String> properties = taskInfo.getProvisionerProperties();
+    ProvisionerContext context = new DefaultProvisionerContext(taskInfo.getProgramRunId(), properties,
+                                                               sparkCompat, createSSHContext(taskInfo.getSshKeyInfo()));
+    return new DeprovisionTask(taskInfo, provisioner, context, provisionerNotifier,
+                               locationFactory, transactional, datasetFramework, 300, taskCleanup);
+  }
+
+  private List<ProvisioningTaskInfo> getInProgressTasks() {
+    return Retries.callWithRetries(
+      () -> Transactionals.execute(transactional, dsContext -> {
+        ProvisionerDataset provisionerDataset = ProvisionerDataset.get(dsContext, datasetFramework);
+        return provisionerDataset.listTaskInfo();
+      }),
+      co.cask.cdap.common.service.RetryStrategies.exponentialDelay(10, 120, TimeUnit.SECONDS),
+      t -> {
+        // Always retry, but log unexpected types of failures
```

Comment:
The retry should also pay attention to the shutdown of this ProvisioningService. Otherwise if there is failure, it can hang the shutdown of cdap master.